### DISCUSSION TITLE
`./setup test` fails with fork-linux and eventfd-linux after Shadow version v3.0.0 on Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by higher-order** [shadow/shadow#3359] [Q&A] [answered]
I have tried to install shadow on multiple machines. On one machine with Ubuntu 22.04, I can build Shadow from the repo on the main branch and have `./setup test` pass all test cases. On other two machines with Ubuntu 20.04, if I build the main branch, it fails many tests. If I checkout v3.2.0, it fails `fork-linux` and `eventfd-linux`. If I build v3.1.0 or 3.0.0, it fails `eventfd-linux`. I can only build and pass all test cases if I checkout v2.5.0.

I can't upgrade Ubuntu from 20.04 to 22.04 at the moment. Is it better to run Shadow v2.5.0 on those machines, or run v3.2.0 with two failed test cases? What's causing the failures and how will they impact experiments?

I've attached the log file of `./setup test --rerun-failed --verbose` when testing v3.2.0 on Ubuntu 20.04. The `--output-on-failure` argument was not accepted by the script. The `/proc/version` is `Linux version 5.4.0-186-generic (buildd@lcy02-amd64-100) (gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.2)) #206-Ubuntu SMP Fri Apr 26 12:31:10 UTC 2024`.

[shadow-failed-tests.v3.2.0.log](https://github.com/user-attachments/files/15938386/shadow-failed-tests.v3.2.0.log)

**Comment by stevenengler** [comment, marked as answer]
Shadow should build and pass all tests on Ubuntu 20.04 as it's one of the [supported platforms](https://shadow.github.io/docs/guide/supported_platforms.html#officially-supported-platforms). If any tests are failing on the main branch please open an issue, ideally one issue for each test if there are only a small number of failing tests. If there are a lot of failures, then one issue for all tests is probably best.

That being said, the shadow tests are the most important tests. The linux tests are just to make sure shadow is behaving similarly to the kernel. Since the kernel's behaviour may change with new linux releases, the linux tests do occasionally fail on some kernel versions. As long as all of the shadow tests pass then shadow should be behaving as expected. Since we run the linux tests in the CI, we know that they are correct for *some* linux kernel version. But we still want to fix the linux tests so that they pass on all supported platforms.

> Is it better to run Shadow v2.5.0 on those machines, or run v3.2.0 with two failed test cases?

If only the linux tests are failing, I would recommend running v3.2.0.

> What's causing the failures

1. > `Testing test_eventfd_readv_writev_nonblock... ✗ (EINVAL: Invalid argument)`
    
    It's probably a difference between kernel versions. Your kernel version is older than what we use in the CI tests (we can only use the oldest kernel that github gives us, which is the kernel for ubuntu 22.04) My guess is that the behaviour of readv/writev for eventfds probably changed at some point between ubuntu 20.04 and 22.04.

3. > `panic: PanicInfo { payload: Any { .. }, message: Some(called 'Result::unwrap()' on an 'Err' value: Errno::EINVAL), location: Location { file: "test/clone/test_fork.rs", line: 1221, col: 10 }, can_unwind: true, force_no_backtrace: false }`
    
    `CLONE_CLEAR_SIGHAND` was added in Linux 5.5, so the `clone3` call in the test fails for your kernel version of 5.4.

> how will they impact experiments?

1. It's unlikely that applications actually use readv/writev on an eventfd. Those tests are there mostly for completeness. Shadow should still behaving as expected. This shouldn't have any impact on experiments.
2. Applications built on Ubuntu 20.04 should hopefully not be using `CLONE_CLEAR_SIGHAND` since it's not supported by Ubuntu 20.04's kernel. Even if they did, Shadow does support this flag and should behave properly. This shouldn't have any impact on experiments.

**Reply by higher-order** [reply]
Thank you @stevenengler. It makes sense that different kernel versions affect the kernel tests. I ran an experiment with Shadow v3.2.0 and it seems to behave normally.

I remember coming across another post that had many tests failures when running on the main branch. But I can't find it now. If I'm able to find it later I'll link it.
